drivers/google/compute_util.go

```diff
@@ -154,32 +162,60 @@ func (c *ComputeUtil) region() string {
 	return c.zone[:len(c.zone)-2]
 }
 
-func (c *ComputeUtil) firewallRule() (*raw.Firewall, error) {
-	return c.service.Firewalls.Get(c.project, firewallRule).Do()
+func (c *ComputeUtil) externalFirewallRule() (*raw.Firewall, error) {
+	return c.service.Firewalls.Get(c.project, c.externalFirewallRuleName()).Do()
 }
 
-func missingOpenedPorts(rule *raw.Firewall, ports []string) map[string][]string {
-	missing := map[string][]string{}
-	opened := map[string]bool{}
+func (c *ComputeUtil) internalFirewallRule() (*raw.Firewall, error) {
+	return c.service.Firewalls.Get(c.project, c.internalFirewallRuleName()).Do()
+}
 
+func (c *ComputeUtil) externalFirewallRuleName() string {
+	return name.SafeConcatName(c.externalFirewallRulePrefix, externalFirewallRuleSuffix)
+}
+
+func (c *ComputeUtil) internalFirewallRuleName() string {
+	return name.SafeConcatName(c.internalFirewallRulePrefix, internalFirewallRuleSuffix)
+}
+
+// updatePorts compares the provided firewall rule against the list of provided ports
+// and returns a boolean indicating if the provided rule has been updated to only include
+// the provided ports.
+func updatePorts(rule *raw.Firewall, ports []string) bool {
+	requestedPorts := map[string][]string{}
+	for _, p := range ports {
+		port, proto := driverutil.SplitPortProto(p)
+		requestedPorts[proto] = append(requestedPorts[proto], port)
+	}
+
+	opened := map[string][]string{}
 	for _, allowed := range rule.Allowed {
 		for _, allowedPort := range allowed.Ports {
-			opened[allowedPort+"/"+allowed.IPProtocol] = true
+			opened[allowed.IPProtocol] = append(opened[allowed.IPProtocol], allowedPort)
 		}
 	}
 
-	for _, p := range ports {
-		port, proto := driverutil.SplitPortProto(p)
-		if !opened[port+"/"+proto] {
-			missing[proto] = append(missing[proto], port)
-		}
+	// if all the ports match then there is no need
+	// to update / recreate the rule.
+	if reflect.DeepEqual(requestedPorts, opened) {
```

Comment:
I think `reflect.DeepEqual` returns false when two slices have the same elements in a different order.